subsys/emds/emds.c

```diff
@@ -6,55 +6,332 @@
+					ch->entry.id, ch->entry.data, ch->entry.len);
+		if (len == ch->entry.len) {
+			LOG_DBG("Write dynamic entry: %d, Data length: %d",
+				ch->entry.id, ch->entry.len);
+			LOG_HEXDUMP_DBG(ch->entry.data, ch->entry.len,
+					"Preserved data: ");
+		} else {
+			LOG_ERR("Write dynamic entry failed (%d:%d).\n",
+				ch->entry.len, len);
+		}
+	}
+
+	emds_store_in_progress = false;
+	k_sched_unlock();
+}
+
+K_THREAD_DEFINE(emds_tid, EMDS_STACK_SIZE, emds_handler, NULL, NULL, NULL,
+		EMDS_PRIORITY, 0, 0);
+
+static int emds_fs_init(void)
+{
+	int rc;
+	const struct flash_area *fa;
+	uint32_t sector_cnt = 1;
+	struct flash_sector hw_fs;
+	size_t emds_sector_size;
+	size_t emds_storage_size = 0;
+	uint16_t cnt = 0;
+
+	rc = flash_area_open(FLASH_AREA_ID(emds_storage), &fa);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_area_get_sectors(FLASH_AREA_ID(emds_storage), &sector_cnt,
+						  &hw_fs);
+	if (rc == -ENODEV) {
+		return rc;
+	} else if (rc != 0 && rc != -ENOMEM) {
+		k_panic();
+	}
+
+	emds_sector_size = hw_fs.fs_size;
+
+	if (emds_sector_size > UINT16_MAX) {
+		return -EDOM;
+	}
+
+	while (cnt < CONFIG_EMDS_SECTOR_COUNT) {
+		emds_storage_size += emds_sector_size;
+		if (emds_storage_size > fa->fa_size) {
+			break;
+		}
+		cnt++;
+	}
+
+	default_emds_storage.cf_emds.sector_size = emds_sector_size;
+	default_emds_storage.cf_emds.sector_count = cnt;
+	default_emds_storage.cf_emds.offset = fa->fa_off;
+	default_emds_storage.flash_dev_name = fa->fa_dev_name;
+
+	rc = emds_flash_init(&default_emds_storage.cf_emds, fa->fa_dev_name);
+	if (rc) {
+		return rc;
+	}
+
+	return 0;
+}
+
+
+static int emds_entries_size(uint32_t *size)
+{
+	int entries = 0;
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		*size += ch->len;
+		entries++;
+	}
+
+	struct emds_dynamic_entry *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		*size += ch->entry.len;
+		entries++;
+	}
+
+	return entries;
+}
+
 int emds_init(void)
 {
+	static bool initialized;
+	int rc;
+
+	if (initialized) {
+		LOG_DBG("Already initialized");
+		return -EALREADY;
+	}
+
+	rc = emds_fs_init();
+	if (rc) {
+		return rc;
+	}
+
+	sys_slist_init(&emds_dynamic_entries);
+
+	initialized = true;
+
+	LOG_DBG("Emergency Data Storage initialized.\n");
+
 	return 0;
 }
 
-int emds_entry_add(struct emds_dynamic_entry *entry)
+int emds_entry_add(struct emds_entry *entry)
 {
+	struct emds_dynamic_entry *ch;
+	struct emds_dynamic_entry *item;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		if (ch->entry.id == entry->id) {
+			return -EINVAL;
+		}
+	}
+
+	item = k_malloc(sizeof(struct emds_dynamic_entry));
+	if (!item) {
+		return -ENOSPC;
+	}
+
+	item->entry = *entry;
+	sys_slist_append(&emds_dynamic_entries, &item->node);
+
 	return 0;
 }
 
 int emds_store(void)
 {
+	if (!emds_ready) {
+		return -ECANCELED;
+	}
+
+	/* Lock all interrupts */
+	(void)irq_lock();
+	/* Stop the radio */
+	nrf_radio_task_trigger(NRF_RADIO, NRF_RADIO_TASK_DISABLE);
```

Comment:
There should not be a link to nrf in this module. I will remove the stopping of the radio. This can be put in the application instead, and other power saving measures if the user want. We can write this in the doc and show in a example instead.